docs/developers/style-guide.md

```diff
@@ -2,6 +2,94 @@
+
+Names should be as self-explanatory as possible, within reason.
+This includes python identifiers (classes, variables, methods, modules, etc.), file names and configuration parameters.
+For example:
+
+* Avoid abbreviations, especially domain-specific ones. Ex. `bs -> batch_size`.
+* Avoid meaningless variable names such as `x`, except for generic methods.
+* Avoid redundancies especially for configuration parameters, ex. `data.data_type` -> `data.type`.
+* Avoid name parts that refer to the data type, ex. `num`. Use type hints instead.
+
+Note that these conventions are enforced more strictly on user-facing names since they are more difficult to change,
+for example configuration parameters and the public interface of core classes and modules.
+
+## 🛬 Imports
+
+We use the following conventions for imports (other than those enforced by isort):
+
+* Import standard library and third party modules by module (ex. `import package.module`, not `from package.module import method`).
+This keeps identifier's origin explicit.
+* Avoid renaming with `as`, except for some (arbitrarily chosen) common ones: `numpy as np`, `triton.language as tl`.
+* Import first-party modules through specific identifiers (ex. `from fast_llm.module import method`, not `import fast_llm.module`). This facilitates the tracking of where those objects are used.
+* Always use absolute imports (ex. no `from .module import method`)
+* Include all explicitly-imported third-party module to `setup.cgf`.
+Only add new requirements if they provide a substantial benefit,
+as we try to keep the requirements to a minimum.
+* Prefer file-level imports over imports inside methods, unless they significantly slow down the import process
+or concern an optional dependency that should not be absolutely required to import the module (ex. `transformers`).
+If an offending import is only required for a type hint, include it in a `if typing.TYPE_CHECKING:` block.
+
+!!! warning "Configuration modules"
+
+    Fast-LLM supports instantation and validation of configurations with a barebone installation.
+    Because of this, modules that contain configuration classes (usually names `config.py`)
+    should not include any top-level third-party import (except for those installed in the [barebone install](https://github.com/ServiceNow/Fast-LLM/blob/main/setup.cfg)),
+    and the same applies for configuration initialization and validation methods.
+    Third-party import may be included in other methods if needed.
+
+## 🔓 Public and Private Interface
+
+Although good practices of object-oriented programming are generally ignored in python,
+Fast-LLM attempts to follow them to an extant, while avoiding unnecessary bloat:
+
+* Mark private and protected variables with an underscore `_` prefix.
+As is customary in python, we make no distinction between the two and avoid the double-underscore `__` notation.
+* Keep public interfaces (methods and variables without underscore prefix) as lean as possible.
```

Comment:
I think it would help if it was explained/exemplified what should remain public vs. private